assets/javascripts/discourse/components/chat-message-collapser.js

```diff
@@ -1,78 +1,125 @@
-        return this.uploads[0].original_filename;
+  @computed("cooked")
+  get youtubeCooked() {
+    const elements = Array.prototype.slice.call(domFromString(this.cooked));
+
+    return elements.reduce((acc, e) => {
+      if (youtubePredicate(e)) {
+        const id = e.dataset.youtubeId;
+        const link = `https://www.youtube.com/watch?v=${escape(id)}`;
+        const title = e.dataset.youtubeTitle;
+        const header = htmlSafe(
+          `<a target="_blank" class="chat-message-collapser-link" rel="noopener noreferrer" href="${link}">${title}</a>`
+        );
+        acc.push({ header, body: e, needsCollapser: true });
       } else {
-        return I18n.t(`chat.uploaded_files`, { count: this.uploads.length });
+        acc.push({ body: e, needsCollapser: false });
       }
-    }
+      return acc;
+    }, []);
   },
 
-  @computed("cooked")
-  get youtubeTitle() {
-    return domFromString(this.cooked).dataset.youtubeTitle;
+  @computed("uploads")
+  get uploadsHeader() {
+    if (this.uploads.length === 1) {
+      return this.uploads[0].original_filename;
+    } else {
+      return I18n.t(`chat.uploaded_files`, { count: this.uploads.length });
+    }
   },
 
   @computed("cooked")
-  get youtubeLink() {
-    const id = domFromString(this.cooked).dataset.youtubeId;
-    return `https://www.youtube.com/watch?v=${escape(id)}`;
-  },
-
-  @action
-  open() {
-    this.set("collapsed", false);
-  },
+  get imageOneboxCooked() {
+    const elements = Array.prototype.slice.call(domFromString(this.cooked));
 
-  @action
-  close() {
-    this.set("collapsed", true);
+    return elements.reduce((acc, e) => {
+      if (imageOneboxPredicate(e)) {
+        const link = animatedImagePredicate(e)
+          ? e.firstChild.src
+          : e.firstElementChild.href;
+        const header = htmlSafe(
+          `<a target="_blank" class="chat-message-collapser-link-small" rel="noopener noreferrer" href="${link}">${link}</a>`
+        );
+        acc.push({ header, body: e, needsCollapser: true });
+      } else {
+        acc.push({ body: e, needsCollapser: false });
+      }
+      return acc;
+    }, []);
   },
 
   @computed("cooked")
-  get isYoutube() {
-    return isYoutube(this.cooked);
+  get hasYoutube() {
+    return hasYoutube(this.cooked);
   },
 
-  @computed("uploads", "cooked")
-  get isImage() {
-    return isImage(this.uploads, this.cooked);
+  @computed("uploads")
+  get hasUploads() {
+    return hasUploads(this.uploads);
   },
 
-  @computed("cooked", "message")
-  get isAnimatedImage() {
-    return isAnimatedImage(this.cooked, this.message);
+  @computed("cooked")
+  get hasImageOnebox() {
+    return hasImageOnebox(this.cooked);
   },
 });
 
-function isYoutube(cooked) {
-  return /^<div class="onebox lazyYT lazyYT-container"/.test(cooked);
+function youtubePredicate(e) {
+  return (
+    e.classList.length &&
+    e.classList.contains("onebox") &&
+    e.classList.contains("lazyYT-container")
```

Comment:
Favouring this over `e.className === "onebox lazyYT..."` in case of changes in the order of classes.